File: benchmarks/StringThing.Benchmarks/EndToEndBenchmarks.cs
using BenchmarkDotNet.Attributes;
using Dapper;
using Microsoft.Data.Sqlite;
using StringThing.Sqlite.Dapper;

namespace StringThing.Benchmarks;

[MemoryDiagnoser]
public class EndToEndBenchmarks
{
    private SqliteConnection _connection = null!;

    [GlobalSetup]
    public void Setup()
    {
        _connection = new SqliteConnection("Data Source=:memory:");
        _connection.Open();

        using var create = _connection.CreateCommand();
        create.CommandText = """
            CREATE TABLE users (
                id INTEGER PRIMARY KEY,
                name TEXT NOT NULL,
                email TEXT,
                age INTEGER NOT NULL,
                active INTEGER NOT NULL
            )
            """;
        create.ExecuteNonQuery();

        for (var i = 1; i <= 100; i++)
        {
            using var insert = _connection.CreateCommand();
            insert.CommandText = $"INSERT INTO users (id, name, email, age, active) VALUES ({i}, 'user_{i}', 'user_{i}@example.com', {20 + i % 50}, {(i % 3 == 0 ? 0 : 1)})";
            insert.ExecuteNonQuery();
        }
    }

    [GlobalCleanup]
    public void Cleanup()
    {
        _connection.Dispose();
    }

    public record User(long Id, string Name, string? Email, long Age, long Active);

    // --- Single row query: 1 parameter ---

    [Benchmark(Description = "Dapper: QuerySingle 1 param")]
    public User Dapper_QuerySingle_OneParam()
    {
        return _connection.QuerySingle<User>(
            "SELECT id, name, email, age, active FROM users WHERE id = @userId",
            new { userId = 42 });
    }

    [Benchmark(Description = "StringThing: QueryStringSingle 1 param")]
    public User StringThing_QuerySingle_OneParam()
    {
        var userId = 42;
        return _connection.QueryStringSingle<User>(
            $"SELECT id, name, email, age, active FROM users WHERE id = {userId}");
    }

    // --- Multiple rows query: 2 parameters ---

    [Benchmark(Description = "Dapper: Query 2 params")]
    public List<User> Dapper_Query_TwoParams()
    {
        return _connection.Query<User>(
            "SELECT id, name, email, age, active FROM users WHERE age > @minAge AND active = @active ORDER BY id",
            new { minAge = 30, active = 1 }).ToList();
    }

    [Benchmark(Description = "StringThing: QueryString 2 params")]
    public List<User> StringThing_Query_TwoParams()
    {
        var minAge = 30;
        var active = 1;
        return _connection.QueryString<User>(
            $"SELECT id, name, email, age, active FROM users WHERE age > {minAge} AND active = {active} ORDER BY id").ToList();
    }

    // --- Query with 5 parameters ---

    [Benchmark(Description = "Dapper: Query 5 params")]
    public List<User> Dapper_Query_FiveParams()
    {
        return _connection.Query<User>(
            "SELECT id, name, email, age, active FROM users WHERE id > @minId AND id < @maxId AND age >= @minAge AND age <= @maxAge AND active = @active",
            new { minId = 10, maxId = 50, minAge = 25, maxAge = 40, active = 1 }).ToList();
    }

    [Benchmark(Description = "StringThing: QueryString 5 params")]
    public List<User> StringThing_Query_FiveParams()
    {
        var minId = 10;
        var maxId = 50;
        var minAge = 25;
        var maxAge = 40;
        var active = 1;
        return _connection.QueryString<User>(
            $"SELECT id, name, email, age, active FROM users WHERE id > {minId} AND id < {maxId} AND age >= {minAge} AND age <= {maxAge} AND active = {active}").ToList();
    }

    // --- Execute (insert) ---

    private int _insertCounter = 1000;

    [Benchmark(Description = "Dapper: Execute insert")]
    public int Dapper_Execute_Insert()
    {
        var id = Interlocked.Increment(ref _insertCounter);
        return _connection.Execute(
            "INSERT OR IGNORE INTO users (id, name, email, age, active) VALUES (@id, @name, @email, @age, @active)",
            new { id, name = "bench", email = "bench@test.com", age = 30, active = 1 });
    }

    [Benchmark(Description = "StringThing: ExecuteString insert")]
    public int StringThing_Execute_Insert()
    {
        var id = Interlocked.Increment(ref _insertCounter);
        var name = "bench";
        var email = "bench@test.com";
        var age = 30;
        var active = 1;
        return _connection.ExecuteString(
            $"INSERT OR IGNORE INTO users (id, name, email, age, active) VALUES ({id}, {name}, {email}, {age}, {active})");
    }

    // --- IN list expansion ---

    [Benchmark(Description = "Dapper: IN @list 10 items")]
    public List<User> Dapper_InList_Ten()
    {
        var ids = Enumerable.Range(1, 10).ToList();
        return _connection.Query<User>(
            "SELECT id, name, email, age, active FROM users WHERE id IN @ids",
            new { ids }).ToList();
    }

    [Benchmark(Description = "StringThing: InList 10 items")]
    public List<User> StringThing_InList_Ten()
    {
        var ids = Enumerable.Range(1, 10).ToArray();
        return _connection.QueryString<User>(
            $"SELECT id, name, email, age, active FROM users WHERE id IN {StringThing.Sqlite.Sqlite.InList([.. ids])}").ToList();
    }

    [Benchmark(Description = "Dapper: IN @list 100 items")]
    public List<User> Dapper_InList_Hundred()
    {
        var ids = Enumerable.Range(1, 100).ToList();
        return _connection.Query<User>(
            "SELECT id, name, email, age, active FROM users WHERE id IN @ids",
            new { ids }).ToList();
    }

    [Benchmark(Description = "StringThing: InList 100 items")]
    public List<User> StringThing_InList_Hundred()
    {
        var ids = Enumerable.Range(1, 100).ToArray();
        return _connection.QueryString<User>(
            $"SELECT id, name, email, age, active FROM users WHERE id IN {StringThing.Sqlite.Sqlite.InList([.. ids])}").ToList();
    }
}

File: fsharp/benchmarks/StringThing.FSharp.Benchmarks/EndToEndBenchmarks.fs
namespace StringThing.FSharp.Benchmarks

open System.Threading
open BenchmarkDotNet.Attributes
open Microsoft.Data.Sqlite
open StringThing.FSharp
open StringThing.FSharp.Sqlite

type BenchmarkUser =
    { Id: int64
      Name: string
      Email: string option
      Age: int64
      Active: int64 }

module BenchmarkUser =
    let row : RowReader<BenchmarkUser> =
        row {
            let! id     = Row.int64 "id"
            and! name   = Row.string "name"
            and! email  = Row.stringOption "email"
            and! age    = Row.int64 "age"
            and! active = Row.int64 "active"
            return { Id = id; Name = name; Email = email; Age = age; Active = active }
        }

[<MemoryDiagnoser>]
type EndToEndBenchmarks() =

    let mutable connection : SqliteConnection = Unchecked.defaultof<_>
    let mutable insertCounter = 1000

    [<GlobalSetup>]
    member _.Setup() =
        connection <- new SqliteConnection("Data Source=:memory:")
        connection.Open()

        use create = connection.CreateCommand()
        create.CommandText <-
            "CREATE TABLE users ( \
                id INTEGER PRIMARY KEY, \
                name TEXT NOT NULL, \
                email TEXT, \
                age INTEGER NOT NULL, \
                active INTEGER NOT NULL)"
        create.ExecuteNonQuery() |> ignore

        for i = 1 to 100 do
            use insert = connection.CreateCommand()
            insert.CommandText <-
                sprintf "INSERT INTO users (id, name, email, age, active) VALUES (%d, 'user_%d', 'user_%d@example.com', %d, %d)"
                    i i i (20 + i % 50) (if i % 3 = 0 then 0 else 1)
            insert.ExecuteNonQuery() |> ignore

    [<GlobalCleanup>]
    member _.Cleanup() =
        connection.Dispose()

    // --- Zero-parameter scalar ---

    [<Benchmark(Description = "Raw: ExecuteScalar 0 param")>]
    member _.Raw_ExecuteScalar_0Param() : int64 =
        use cmd = connection.CreateCommand()
        cmd.CommandText <- "SELECT COUNT(*) FROM users"
        cmd.ExecuteScalar() :?> int64

    [<Benchmark(Description = "FSharp: ExecuteStringScalar 0 param")>]
    member _.FSharp_ExecuteScalar_0Param() : int64 =
        let result =
            connection.ExecuteStringScalar $"""
                SELECT COUNT(*) FROM users
                """
        match result with
        | Some v -> v :?> int64
        | None -> 0L

    // --- QuerySingle 1 param ---

    [<Benchmark(Description = "Raw: QuerySingle 1 param")>]
    member _.Raw_QuerySingle_1Param() : BenchmarkUser =
        use cmd = connection.CreateCommand()
        cmd.CommandText <- "SELECT id, name, email, age, active FROM users WHERE id = @userId"
        cmd.Parameters.AddWithValue("@userId", 42) |> ignore
        use reader = cmd.ExecuteReader()
        reader.Read() |> ignore
        { Id = reader.GetInt64(0)
          Name = reader.GetString(1)
          Email = if reader.IsDBNull(2) then None else Some (reader.GetString(2))
          Age = reader.GetInt64(3)
          Active = reader.GetInt64(4) }

    [<Benchmark(Description = "FSharp: QueryStringSingle 1 param")>]
    member _.FSharp_QuerySingle_1Param() : BenchmarkUser =
        let userId = 42L
        connection.QueryStringSingle BenchmarkUser.row $"""
            SELECT id, name, email, age, active
            FROM users
            WHERE id = {userId}
            """

    // --- QuerySingle 2 params ---

    [<Benchmark(Description = "Raw: Query 2 params")>]
    member _.Raw_Query_2Params() : BenchmarkUser =
        use cmd = connection.CreateCommand()
        cmd.CommandText <-
            "SELECT id, name, email, age, active FROM users \
             WHERE age > @minAge AND active = @active \
             ORDER BY id LIMIT 1"
        cmd.Parameters.AddWithValue("@minAge", 30) |> ignore
        cmd.Parameters.AddWithValue("@active", 1) |> ignore
        use reader = cmd.ExecuteReader()
        reader.Read() |> ignore
        { Id = reader.GetInt64(0)
          Name = reader.GetString(1)
          Email = if reader.IsDBNull(2) then None else Some (reader.GetString(2))
          Age = reader.GetInt64(3)
          Active = reader.GetInt64(4) }

    [<Benchmark(Description = "FSharp: QueryString 2 params")>]
    member _.FSharp_Query_2Params() : BenchmarkUser =
        let minAge = 30L
        let active = 1L
        connection.QueryStringSingle BenchmarkUser.row $"""
            SELECT id, name, email, age, active
            FROM users
            WHERE age > {minAge} AND active = {active}
            ORDER BY id
            LIMIT 1
            """

    // --- QuerySingle 5 params ---

    [<Benchmark(Description = "Raw: Query 5 params")>]
    member _.Raw_Query_5Params() : BenchmarkUser =
        use cmd = connection.CreateCommand()
        cmd.CommandText <-
            "SELECT id, name, email, age, active FROM users \
             WHERE id > @minId AND id < @maxId AND age >= @minAge AND age <= @maxAge AND active = @active \
             LIMIT 1"
        cmd.Parameters.AddWithValue("@minId", 10) |> ignore
        cmd.Parameters.AddWithValue("@maxId", 50) |> ignore
        cmd.Parameters.AddWithValue("@minAge", 25) |> ignore
        cmd.Parameters.AddWithValue("@maxAge", 40) |> ignore
        cmd.Parameters.AddWithValue("@active", 1) |> ignore
        use reader = cmd.ExecuteReader()
        reader.Read() |> ignore
        { Id = reader.GetInt64(0)
          Name = reader.GetString(1)
          Email = if reader.IsDBNull(2) then None else Some (reader.GetString(2))
          Age = reader.GetInt64(3)
          Active = reader.GetInt64(4) }

    [<Benchmark(Description = "FSharp: QueryString 5 params")>]
    member _.FSharp_Query_5Params() : BenchmarkUser =
        let minId = 10L
        let maxId = 50L
        let minAge = 25L
        let maxAge = 40L
        let active = 1L
        connection.QueryStringSingle BenchmarkUser.row $"""
            SELECT id, name, email, age, active
            FROM users
            WHERE id > {minId}
              AND id < {maxId}
              AND age >= {minAge}
              AND age <= {maxAge}
              AND active = {active}
            LIMIT 1
            """

    // --- Execute insert ---

    [<Benchmark(Description = "Raw: Execute insert")>]
    member _.Raw_Execute_Insert() : int =
        let id = Interlocked.Increment(&insertCounter)
        use cmd = connection.CreateCommand()
        cmd.CommandText <-
            "INSERT OR IGNORE INTO users (id, name, email, age, active) VALUES (@id, @name, @email, @age, @active)"
        cmd.Parameters.AddWithValue("@id", id) |> ignore
        cmd.Parameters.AddWithValue("@name", "bench") |> ignore
        cmd.Parameters.AddWithValue("@email", "bench@test.com") |> ignore
        cmd.Parameters.AddWithValue("@age", 30) |> ignore
        cmd.Parameters.AddWithValue("@active", 1) |> ignore
        cmd.ExecuteNonQuery()

    [<Benchmark(Description = "FSharp: ExecuteString insert")>]
    member _.FSharp_Execute_Insert() : int =
        let id = int64 (Interlocked.Increment(&insertCounter))
        let name = "bench"
        let email = "bench@test.com"
        let age = 30L
        let active = 1L
        connection.ExecuteString $"""
            INSERT OR IGNORE INTO users (id, name, email, age, active)
            VALUES ({id}, {name}, {email}, {age}, {active})
            """
